doc/1.4/language.md

```diff
@@ -3249,6 +3249,176 @@ local (int i, uint8 j) = m(e1);
+    }
+
+    method write_register(uint64 written, uint64 enabled_bytes,
+                          void *aux) default {
+        if (write_allowed()) {
+            default(written, enabled_bytes, aux);
+        } else {
+            on_write_attempted_when_not_allowed();
+        }
+    }
+}
+
+template very_complex_gated_register is (very_complex_register,
+                                         gated_register) {
+    // No sensible way to combine the two implementations by calling both.
+    // Even if there were, calling both implementations would cause each field
+    // of the register to be written to multiple times, potentially duplicating
+    // side-effects, which is undesirable.
+    // Instead, very_complex_register is chosen as the base implementation
+    // called, and the behaviour of gated_register is replicated around that
+    // call.
+    method write_register(uint64 written, uint64 enabled_bytes,
+                          void *aux) default {
+        if (write_allowed()) {
+            this.templates.very_complex_register.write_register(
+                written, enabled_bytes, aux);
+        } else {
+            on_write_attempted_when_not_allowed();
+        }
+    }
+}
+
+in each (gated_register, very_complex_register) {
+    is very_complex_gated_register;
+}
+```
+
+A template-qualified method implementation call is resolved by using
+the method implementation provided to the object by the named template.
+If no such implementation is provided (whether it be because the template does
+not specify one, or specifies one which is not provided to the object due to its
+definition being eliminated by an [`#if`](#conditional-objects)), then the
+ancestor templates of the named template are recursively searched for the
+highest-rank (most specific) implementation provided by them. If the ancestor
+templates provide multiple hierarchically unrelated implementations, then the
+choice is ambiguous and the call will be rejected by the compiler. In this case,
+the modeller must refine the template-qualified method implementation call to
+name the ancestor template whose implementation they would like to use.
+
+A template-qualified method implementation call done via [a value of template
+type](#templates-as-types) functions differently compared to compile-time
+object references. In particular, `this.templates` within the bodies of `shared`
+methods functions differently. The specified template must be an ancestor
+template of the value's template type, the <tt>object</tt> template, or the
+template type itself; furthermore, the specified template **must provide or
+inherit a `shared` implementation of the named method**. It is not sufficient
+that the method is simply _declared_ `shared` such that it is part of the
+template type: the implementation itself must also be `shared`. For more
+information, see the documentation of the [`ENSHAREDTQMIC` error
+message.](messages.html#ENSHAREDTQMIC)
+
```

Comment:
This diff is 2 things: I moved the TQMIC subsubsection to the Method Statements subsection (after Method Calls), and I moved out the final part about resolving `ENSHAREDTQMIC` to the documentation of `ENSHAREDTQMIC`, replacing it with a forward reference to that documentation.